### PR TITLE
DocumentOverlays schema and scaffolding

### PR DESCRIPTION
**Pull request by wu-hui**
Creates a new SQLite table "document_overlays" to save overlays for documents that are modified locally. An overlay is just a `Mutation`.

### CODE REVIEW

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java [hunk 1]
```diff
@@ -255,7 +255,8 @@ public static FirebaseFirestore testFirestore(
     // This unfortunately is a global setting that affects existing Firestore clients.
     Logger.setLogLevel(logLevel);
 
-    // TODO: Remove this once this is ready to ship.
+    // TODO: Remove below once this is ready to ship.
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java:258]: Please use "TODO(Overlay)"

File: firebase-firestore/firebase-firestore.gradle [hunk 2]
```diff
@@ -111,6 +111,12 @@ android.libraryVariants.all { variant ->
     } else {
         variant.buildConfigField("boolean", "ENABLE_INDEXING", "false")
     }
+
+    if (localProps['firestoreEnableOverlay']) {
```
Comment (schmidt-sebastian) [inline firebase-firestore/firebase-firestore.gradle:115]: Should we add a TODO to remove this in the future?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlays.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+package com.google.firebase.firestore.local;
+
+import androidx.annotation.Nullable;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.mutation.Mutation;
+
+/**
+ * Provides methods to read and write document overlays.
+ *
+ * <p>An overlay is a saved {@code Mutation}, that gives a local view of a document, when applied to
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlays.java:24]: Remove comma before "when".

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlays.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * the remote version of the document.
+ */
+public interface DocumentOverlays {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlays.java:27]: Classes like this should be in singular form. Java uses plural form for classes with a bunch of static helpers (Utils.java, Helpers.java, Values.java, Collections.java)

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlays.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  /**
+   * Gets the saved overlay mutation for the given user and the given document key. Returns null if
+   * there is no overlay for the user and document combination.
+   */
+  @Nullable
+  Mutation getOverlayMutation(String uid, DocumentKey key);
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlays.java:33]: Please use higher level User object. It makes for a more obvious API (but also see comment about dropping this argument).

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryDocumentOverlays.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+package com.google.firebase.firestore.local;
+
+import androidx.annotation.Nullable;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.mutation.Mutation;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MemoryDocumentOverlays implements DocumentOverlays {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryDocumentOverlays.java:23]: Random naming thought (feel free to ignore): I think `MemoryOverlay` would read nicer.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryDocumentOverlays.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  private Map<String, Map<DocumentKey, Mutation>> overlays = new HashMap<>();
+
+  @Nullable
+  @Override
+  public Mutation getOverlayMutation(String uid, DocumentKey key) {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryDocumentOverlays.java:28]: The way UserIds are handled in the MutationQueue is different. We have a single MutationQueue per uses.  That means that these methods wouldn't need to take the User ID at all.

It also means that you wold not have to implement user switching in this class. Overlays should be destroyed once a user signs out, and not be available once the user signs back in again (to match current Memory-persistence behavior). I noticed that you did not implement User ID switching - so you are already on the right track :)

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/Persistence.java [hunk 8]
```diff
@@ -50,8 +50,12 @@
 public abstract class Persistence {
   static final String TAG = Persistence.class.getSimpleName();
 
+  /** Temporary setting for enabling document overlays. */
+  // TODO(Overlay): Remove this.
+  public static boolean OVERLAY_SUPPORT_ENABLED = BuildConfig.ENABLE_OVERLAY;
+
   /** Temporary setting for enabling indexing-specific code paths while in development. */
-  // TODO: Remove this.
+  // TODO(Indexing): Remove this.
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/Persistence.java:58]: Thanks

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlays.java (hunk 5 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlays.java:33]: Maybe also s/getOverlayMutation/getOverlay (if you do do this, then here and below)

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteDocumentOverlays.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    String path = EncodedPath.encode(key.getPath());
+    return db.query("SELECT overlay_mutation FROM document_overlays WHERE uid = ? AND path = ?")
+        .binding(uid, path)
+        .firstValue(
+            row -> {
+              if (row != null) {
+                try {
+                  return serializer.decodeMutation(row.getBlob(0));
+                } catch (InvalidProtocolBufferException e) {
+                  throw fail("Overlay mutations failed to parse: %s", e);
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteDocumentOverlays.java:45]:
```suggestion
                  throw fail("Overlay failed to parse: %s", e);
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java [hunk 10]
```diff
@@ -51,8 +51,10 @@
    */
   static final int VERSION = 12;
 
+  static final int OVERLAY_SUPPORT_VERSION = VERSION + 1;
+
   // TODO(indexing): Remove this constant and increment VERSION to enable indexing support
-  static final int INDEXING_SUPPORT_VERSION = VERSION + 1;
+  static final int INDEXING_SUPPORT_VERSION = VERSION + 2;
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:57]:
```suggestion
  static final int INDEXING_SUPPORT_VERSION = OVERLAY_SUPPORT_VERSION + 1;
```

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/DocumentOverlaysTestCase.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  @Test
+  public void testCanReadSavedOverlay() {
+    Mutation m = patchMutation("coll/doc1", map("foo", "bar"));
+    overlays.saveOverlayMutation("Alice", key("coll/doc1"), m);
+
+    assertEquals(m, overlays.getOverlayMutation("Alice", key("coll/doc1")));
+  }
+
+  @Test
+  public void testCannotReadSavedOverlayByOtherUser() {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/DocumentOverlaysTestCase.java:87]: Note: This test would not be relevant once we move the User ID handling to Persistence.java

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlay.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Provides methods to read and write document overlays.
+ *
+ * <p>An overlay is a saved {@code Mutation}, that gives a local view of a document when applied to
+ * the remote version of the document.
+ */
+public interface DocumentOverlay {
+  /**
+   * Gets the saved overlay mutation for the given user and the given document key. Returns null if
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/DocumentOverlay.java:29]: Needs update (remove mention of user)

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalSerializer.java [hunk 13]
```diff
@@ -326,4 +327,13 @@ public FieldIndex decodeFieldIndex(
...
   }
+
+  public Mutation decodeMutation(byte[] blob) throws InvalidProtocolBufferException {
+    Write mutation = Write.parseFrom(blob);
+    return rpcSerializer.decodeMutation(mutation);
+  }
+
+  public byte[] encodeMutation(Mutation mutation) {
+    return rpcSerializer.encodeMutation(mutation).toByteArray();
+  }
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalSerializer.java:338]: I just scrolled through the file and noticed that none of the other methods here encode and decode to and from bytes. Rather, this is done at the callsite. This actually has some benefit as this "detaches" this encoder from the underlying storage layer. While SQLite uses `byte[]`, the memory implementation uses the Proto directly. With that, is it possible to convert to bytes in the SQLite implementation of the document overlay?

**Review by schmidt-sebastian — APPROVED** [review]
Thanks for addressing the feedback. Some final thoughts, but looks good.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalSerializer.java (hunk 13 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalSerializer.java]: Makes sense.

**Pull request merged**